polonius-parser/src/token.rs

```diff
@@ -0,0 +1,176 @@
+use std::{
+    fmt,
+    ops::{Index, Range},
+};
+
+#[derive(PartialEq, Eq, Clone, Copy)]
+pub struct Token {
+    pub kind: TokenKind,
+    pub span: Span,
+}
+
+impl fmt::Debug for Token {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "{:?} - <{}, {}>",
+            self.kind, self.span.start, self.span.end
+        )
+    }
+}
+
+impl fmt::Display for Token {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", self.kind)
+    }
+}
+
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+#[repr(u16)]
+pub enum TokenKind {
+    Comma,
+    Colon,
+    Semi,
+    Slash,
+    LParen,
+    RParen,
+    LCurly,
+    RCurly,
+    // relation keywords
+    KwUseOfVarDerefsOrigin,
+    KwDropOfVarDerefsOrigin,
+    KwPlaceholders,
+    KwKnownSubsets,
+    // CFG keywords
+    KwBlock,
+    KwGoto,
+    // effect keywords - facts
+    KwOutlives,
+    KwLoanIssuedAt,
+    KwLoanInvalidatedAt,
+    KwLoanKilledAt,
+    KwVarUsedAt,
+    KwVarDefinedAt,
+    KwOriginLiveOnEntry,
+    KwVarDroppedAt,
+    // effect keywords - use
+    KwUse,
+    // parameters
+    Origin,
+    Block,
+    Loan,
+    Variable,
+    Comment,
+    Whitespace,
+    Error,
+    Eof,
+}
+
+#[macro_export]
+macro_rules! T {
```

Comment:
`TokenKind` is just a `u16` and `Copy`, so there isn't really a need to intern it. The macro is more for convenience, and I also find it more readable since you have a lot of checks against kinds in `match`es and calls like `self.consume(T!['('])?;` or lists like
```rs
ParseError::UnexpectedToken {
    found,
    expected: vec![T![;], T![/]],
    position: self.position(),
}
```
A macro like this is used by the folks over at rust-analyzer, and has come in handy for me in my own projects as well a few times. I've added some explanation in the code that reflects what I'm explaining here.